### PR TITLE
Fedasync/dev

### PR DESCRIPTION
**Pull request by Zilinghan**
Add two async FL algorithms using MPI: FedAsync and FedBuf together with the example code to run those algorithms on partitioned MNIST dataset.

### CODE REVIEW

**Review by kibaekkim — CHANGES_REQUESTED** [review]
Can you add a test script to `tests` directory? See `tests/test_mnist.py`.

**Comment by Zilinghan** [comment]
> Can you add a test script to `tests` directory? See `tests/test_mnist.py`.

Sure, let me see and write one.

**Comment by Zilinghan** [comment]
I have added tests for both FedAsync and FedBuf inside `tests/test_mnist.py` file.

File: src/appfl/run_mpi_async.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+                global_model = server.model.state_dict()
+                gloabl_model_buffer = io.BytesIO()
+                torch.save(global_model, gloabl_model_buffer)
+                global_model_bytes = gloabl_model_buffer.getvalue()
+
+                # Send (buffer size, finish flag) - INFO - to the client in a blocking way
+                comm.send((len(global_model_bytes), False), dest=client_idx+1, tag=client_idx+1)
+
+                # Send the buffered model - MODEL - to the client in a NON-blocking way
+                comm.Send(np.frombuffer(global_model_bytes, dtype=np.byte), dest=client_idx+1, tag=client_idx+1+comm_size)
```
Comment (kibaekkim) [inline src/appfl/run_mpi_async.py:173]: Is this supposed to be `Isend`?

Comment (Zilinghan) [reply src/appfl/run_mpi_async.py:173]: Thanks for this catch! Let me double check this part

Comment (Zilinghan) [reply src/appfl/run_mpi_async.py:173]: I double check the code and find there is a typo in the comment. It should be a blocking send.

File: src/appfl/algorithm/server_fed_buffer.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    def update_gradient(self, local_gradient: dict, init_step: int, client_idx: int):
+        for name, _ in self.model.named_parameters():
+            alpha_t = self.alpha * self.staleness(self.global_step - init_step)
+            self.pseudo_grad[name] += local_gradient[name]  * self.weights[client_idx] * alpha_t
+        self.counter += 1
+
+    def update(self, local_gradient: dict, init_step: int, client_idx: int):  
+        self.update_gradient(local_gradient, init_step, client_idx)
+        if self.counter == self.K:
```
Comment (kibaekkim) [inline src/appfl/algorithm/server_fed_buffer.py:40]: Where is `self.K` set?

**Review by kibaekkim — COMMENTED** [review]
The code looks excellent! I have only one comment.

File: src/appfl/algorithm/server_fed_buffer.py (hunk 2 above)
Comment (Zilinghan) [reply src/appfl/algorithm/server_fed_buffer.py:40]: In the example file `examples/mnist_async_mpi`  line 166, we have `cfg.fed.args.K = args.K`, which puts K into the fed.args. 
And when we instantiate the server, we use the following, where K is passed by `**cfg.fed.args`
`server = eval(cfg.fed.servername)(
        weights, 
        copy.deepcopy(model), 
        loss_fn, num_clients, 
        device,
        **cfg.fed.args
    )
`
For ServerFedBuffer, it inherits FedServer, which deals with those keyword arguments by calling `self.__dict__.update(kwargs)`.

**Review by kibaekkim — APPROVED** [review]

**Pull request merged**